Integrate Nix with IPFS (#133)
NixOS/rfcs: rfcs/0133-ipfs.md

---
feature: ipfs
start-date: (fill me in with today's date, YYYY-MM-DD)
author: John Ericsion (@Ericson2314) on behalf of [Obsidian Systems](https://obsidian.systems)
co-authors: (find a buddy later to help out with the RFC)

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)
> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer: ```suggestion
shepherd-team: edolstra, kevincox, gador, @mjoseph-nixpkgs
shepherd-leader: mjoseph-nixpkgs
```

Author: Woo!

related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

> Integrate Nix with IPFS, in phases of increasing sophistication.

Reviewer 2: While I think most of the proposals in this RFC make sense, I think that proposing them all together makes discussion and acceptance difficult. I would recommend that each proposal is split into its own RFC to keep each change atomic. The whole roadmap is very useful to see however I think this roadmap can be maintained outside of the RFC process or as a bunch of draft RFCs that can then be opened and progressed once the dependencies have been accepted.

Author: Well if every step gets its own RFC it will take quite a while!

I am hoping in the first meeting we can chop this into two parts parts, move the second to future work and circle back to it later. Maybe we do that a few times, but I hope the resulting partition ends up being more than one step per RFC :)

Reviewer 2: Yeah, maybe coarse chops are the right approach.

It is important to remember:

1. The time to review an RFC is super-linear with the size.
2. The number of smaller discussions to be resolved is higher for a large RFC.

So while smaller RFCs may have more bureaucratic overhead and organizational overhead it may help actually solidify each component and get it approved. Because trying to understand and discuss the later changes when the assumptions is is based on are mutable provide valuable flexibility but this flexibility also makes it hard to keep discussions concrete and productive.

Author: Yeah right it's an optimization problem with overhead on balkanized discussion on one hand vs superlinearity of review cost w.r.t RFC on the other.

Moving things on first meeting into the "future work" section is a sort of non-balanced bisection algorithm to approximate an optimal solution :).

Author: I agree, but they also are arguably more "invasive" depending on exactly where we draw the line.

Author: Collapsing IPFS threads.

This follows the work done and described in https://github.com/obsidiansystems/ipfs-nix-guide/ .

# Motivation
[motivation]: #motivation

> ## Binary distribution

Reviewer 3: IIUC, the phase one of this RFC is to use IPFS as a decentralized CDN for binary/source distributions. But I do think there still needs to be a benchmark on whether the transport is more efficient than `git fetch` of Git packfiles.

Author: `git fetch` and packfiles can do very wonderful and clever things, but acting as a CDN is not among them. I think this is an apples-and-oranges comparison without a clear answer benchmarking could resolve.

Reviewer 3: One could have the mirrors in the form of Git repos, though this is not as automatic as running the IPFS daemon.

At the very least, for low-resource/remote systems, in addition to improving the download speed, I think we wouldn't want the package download to consume a lot of bandwidth overall? But then again, this optimization issue is implementation detail, and so is out of scope for this RFC.

Currently distributing Nix binaries takes a lot of bandwidth and storage.
This is a barrier to being a Nix user in areas of slower internet --- which includes the vast majority of the world's population at this time.

> This is also a barrier to users running their own caches.

Reviewer 3: How is it addressed in Obsidian that IPFS (by default, I think) duplicates the storage requirement of a package? Both in the output directory `/nix/store/...` and in IPFS own blocks store.

Author: This is not yet addressed. The focus of this RFC is interfaces not implementation details. I think it is fine that is out of scope for now because we can transparently improve it later.

Also Obsidian Systems is just the name of the company I work at. This project doesn't have a "branded" name, because it is meant to become a seamless part of upstream Nix.

Reviewer 3: SGTM

Content-addressing opens up a *huge* design space of solutions to get around such problems.
IPFS explores many of those solutions.

## Source distribution and archival

A goal of the Nix ecosystem is to package software in a way that never bitrots.
Getting in the way of that, however, is the fact source code frequently goes off-line.
The Software Heritage archive is the best in the world, and a natural partner in this effort.

Unfortunately, as https://www.tweag.io/blog/2020-06-18-software-heritage/ describes at the end, a major challenge is the way nix content-addresses software.
First of all, Nix hashes sources in bespoke ways that no other project will adopt.
Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).

We should natively support git file hashing, which Git repos and Software Heritage both support.

> This will completely obliterate these issues.

Author: > Or does the lack of adoption of SHA2 in Git pose issues.

I think it does. I am happy to prioritize adding SHA-2 Git (and SWHID) content-addressing ahead of wide adoption, and indeed try to lead that change, but I am hesitant to get rid of SHA-1 support until that change happens.

Put another way, as long as Git SHA-1 is omnipresent, our our additional hashes adding much value? Whoever wrote down the additional hashes was verifying something by SHA-1, so a hash collision attack could sneak in then.

> This will completely obliterate these issues.

Author: > Or does the lack of adoption of SHA2 in Git pose issues.

I think it does. I am happy to prioritize adding SHA-2 Git (and SWHID) content-addressing ahead of wide adoption, and indeed try to lead that change, but I am hesitant to get rid of SHA-1 support until that change happens.

Put another way, as long as Git SHA-1 is omnipresent, our our additional hashes adding much value? Whoever wrote down the additional hashes was verifying something by SHA-1, so a hash collision attack could sneak in then.

> This will completely obliterate these issues.

Reviewer 2: I think I am partly confused because it seems like there may be multiple uses of the hash, both for the native importing of sources that use Git (which the sha1 support is definitely a huge benefit) and use for non-git sources and other artifacts (where the benefits are less clear). I think if we can make it clear which hash is used for which things and where parity matters then I think you are right and we will need to support sha1-git for the time being.

> This will completely obliterate these issues.

Author: We could I suppose *only* allow git sha-1 fetching with `builtins.fetchgit` and friends, but what about fixed-output derivations like Nixpkgs uses that also fetch git repos?

One thing I am hoping we can get to is just writing down *one* hash for git repo fetching, and if it is a commit hash or tag object instead of git hash, using a [Merkle inclusion proof](https://computersciencewiki.org/index.php/Merkle_proof) to connect it to the tree hash we'd ultimately use.

That would be a *huge* ergonomics improvement for Nix.

> This will completely obliterate these issues.

Reviewer 2: I'm trying to comprehend the security implications of sha1 hashes for store content and I think they aren't as bad as I thought. Since I think that sha1 hashes in the store will look different (different length) than other hashes you won't be able to replace a non-sha1 path with a colliding sha1 path. So I think it will only reduce security for places where it is used which is probably not that much of a problem. With little thought the worst attack I can think of is forcing someone to pull your source for a build. That is obviously bad but as sha1 is (hopefully) phased out less and less packages will be affected by this, just having sha1 as a supported hash doesn't reduce the integrity of derivations that don't use it.

That being said I wonder if it still makes sense to have `builtins.fetchgit` still support providing an additional hash. This would be less ergonomic than just providing one hash for both commit ID + fixed output verification but would provide higher confidence in the integrity of the downloaded result. So this would basically look like downloading from IPFS via the sha1 git hash and then verifying the result again with the other hash.

> This will completely obliterate these issues.

Author: > So I think it will only reduce security for places where it is used which is probably not that much of a problem.

Yup. And the name of the hash is always included in the store path hash computation too. And the *second* round of hashing used for that is SHA-256.

I should add some text here emphasizing this property.

> This will completely obliterate these issues.

Author: > That being said I wonder if it still makes sense to have `builtins.fetchgit` still support providing an additional hash. This would be less ergonomic than just providing one hash for both commit ID + fixed output verification but would provide higher confidence in the integrity of the downloaded result.

Yeah that is just what we have today, more or less.

> So this would basically look like downloading from IPFS via the sha1 git hash and then verifying the result again with the other hash.

This is a little tricky because right now one *just* fetches from IPFS by using it as a substitutor, this is no `builtins.fetchipfs` because no libfetchers additional logic is needed! But there are still options we can discuss.

> This will completely obliterate these issues.

Reviewer 2: A simple option to work around this would be to keep `fetchgit` as a sha1 derivation but then wrap it in another derivation that just checks the hash and copies the result. You wouldn't be able to download this derivation from IPFS but it is likely cheap enough that it isn't a huge issue to rebuild it locally.

> This will completely obliterate these issues.

Reviewer 4: > Git hashing is currently on SHA1 which is showing weaknesses.

There is some useful and recent information on the git-sha256 situation [in this LWN article](https://lwn.net/Articles/898522/).

> This will completely obliterate these issues.

Author: Thanks! My take away is:

1. Make sure we also support [SHA-1DC](https://github.com/cr-marcstevens/sha1collisiondetection) not vanilla SHA-1.
2. Try out supporting greenfield SHA-256 git repos.
3. Do not worry about migrating code as upstream git doesn't have a plan to migrate / dual-sign repos either.

We didn't do (2) in 2020 when the support in git itself was new, but we could do (2) now.

Ultimately crypto-agility is an interesting and important problem, and I think there is a great opportunity to lead the way here *eventually*, but lets worry about trying to catch up to the state of the art before we push it forward!

> This will completely obliterate these issues.

Reviewer 5: So the git upload is signed by the signing key to make sure that it is trusted?

> This will completely obliterate these issues.

Reviewer 4: > 3\. dual-sign

I think Ericson2314 meant "dual-hash".

> This will completely obliterate these issues.

Reviewer 5: How would dual hashing prevent the following scenario:
- a malicious commit is produced with a sha1 collision
- it gets uploaded to ipfs before the nix CI can process it

Also how would it be prevented that wrongly signed nix store entries are uploaded to ipfs? Would the hash then be different or could a collision be produced?

> This will completely obliterate these issues.

Author: A proper dual hashing solution would be something like `sha256 || sha1` hashes so you have ways to verify not only the root object but every object in it's closure.

> This will completely obliterate these issues.

Author: Collapsing IPFS threads. Since the RFC is now about Git hashing, those drawbacks or more intrinsic to the goal. (Whereas with IPFS we could hash in other way.s)

Each item can be done separately provided its dependent items are also done.

## Augmented `narinfo`

- **Purpose**: Binary distribution

*This is taken from [RFC PR 122](https://github.com/lucasew/rfcs/blob/binary-cache-ipfs/rfcs/0122-binary-cache-ipfs.md), which was abandoned by its author.*

The purpose of this is a "hybrid" store where the narinfo metadata is still severed via HTTPS, but the data itself is served via IPFS.

Today, a narinfo looks like this:

```
StorePath: /nix/store/gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
URL: nar/0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb.nar.xz
Compression: xz
FileHash: sha256:0i6ardx43rdg24ab1nc3mq7f5ykyiamymh1v37gxdv5xh5cm0cmb
FileSize: 40360
NarHash: sha256:1ddv0iqq47j0awyw7a8dmm8bz71c6ifrliq53kmmsfzjxf3rwvb8
NarSize: 197528
References: 7gx4kiv5m0i7d7qkixq2cwzbr10lvxwc-glibc-2.27 gdh8165b7rg4y53v64chjys7mbbw89f9-hello-2.10
Deriver: 5sj6fdfym58sdaf3r5p87v4l8sj2zlvn-hello-2.10.drv
Sig: cache.nixos.org-1:K0thQEG60rzAK8ZS9f1whb7eRlIshlMDJAm7xvX1oF284H+PTqlicv/wGW6BIj+wWWONHvUZ2MYc+KDArekjDA==
```

This RFC proposes new key-value pairs that in this example would be:

```
IpfsCid: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Author: Oh oops, I tried to write down things per the spec.

This is the one part that *isn't* already implemented. And so me writing that has some guesswork.

TBH I feel somewhat demotivated about this field because I am not sure how much UnixFS intends to make writing and verifying it anyways, and unixfs-wrapped NAR is somewhat a stop-gap anyways.

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Reviewer 2: I think the important thing is without this field someone who has a nar can't (without some guesswork and trial-and-error) share it with the network. This means that if we want the official cache.nixos.org to support IPFS it will need to do so natively. By adding this field we add a transition option where cache.nixos.org doesn't need to provide IPFS resources directly but interested community members can "rehost" the nars and others can find them.

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Author: I definitely agree strongly with that on principle. I just hope I don't get very bogged down trying to extra all this metadata and test it. Perhaps the thing to do is only support very few combinations at first.

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Reviewer 2: +1 If we say that the only supported method to start is fixed raw-leaves with 4k chunks or whatever I think that is a good start. I still think the JSON format is nice for future extensibility even if we only currently support one option.

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Author: Yeah we can and should have full JSON metadata if possible even if we only support one case to begin.

> IpfsCid: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN
> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Reviewer 6: Should these perhabs be capitalized as

```suggestion
IPFSCID: Qmf8NfV2hnq44RoQw9vxmSpGYTwAovA8FUCxeCJCqmXeNN
IPFSEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}
```

following the convention that acronym initials are capitalized?

Author: I think that is hard to read, so I did the rust-style "treat acronyms as words for better camel-casing" thing, but I am fine with other options like snake case or kebob case too.

Reviewer 7: i like the rule of acronyms larger than 3 get camel-cased

Author: Collapsing IPFS threads.

> IpfsEncoding: {"method":"wrapped-nar","chunking":{"leaf-format":"raw","strategy":"fixed-size"},"layout":"balanced","max-width":174}

Author: Collapsing IPFS threads.

```

Just as today, the `NarHash` and `NarSize` remain the *normative* way to lock down the store object the `narinfo` file describes.
Conversely, The `URL`, `FileHash` and `FileSize` by contrast are *informational*, describing not what the store object *is*, but *how to get it*.

The `IpfsCid` and `IpfsEncoding` are likewise informational, describing how to get the store object:

- `IpfsCid`: Native content address for IPFS.

> - `IpfsEncoding`: Enough info to deterministically rebuild the IPFS representation from a non-IPFS copy of the store object.

Author: Yes I agree we should prohibit not ignore extra freedom on the axes we don't care about. No Postel's law permissiveness for the first version!

> - `IpfsEncoding`: Enough info to deterministically rebuild the IPFS representation from a non-IPFS copy of the store object.

Author: Collapsing IPFS threads.

   For now, `IpfsEncoding` will only support `unixfs-nar`, which works as follows:

   The NAR is itself wrapped in IPFS's [UnixFS](https://github.com/ipfs/specs/blob/main/UNIXFS.md).
   This other format can be extracted from the CID (which is conceptually a pair of encoding metadata and a hash).
   For now, only IPFS's "unixfs" is supported.
   `chunking`, `layout`, and `max-size` are tuning parameters for unixfs [described in the UnixFS spec](https://github.com/ipfs/specs/blob/main/UNIXFS.md#importing).

>    "UNIXFS" is not used directly because it doesn't support the "executable bit** Nix does on files.

Author: Oh hmm perhaps I am mistaken. @warpfork do you know about this?

>    "UNIXFS" is not used directly because it doesn't support the "executable bit** Nix does on files.

Author: Collapsing IPFS threads.

>    NAR archive are not used directly because IPFS doesn't support arbitrary large objects.

Reviewer 2: ```suggestion
   NAR archive are not used directly because IPFS doesn't support arbitrarily large objects.
```

## IPFS Narinfo and "stateful" IPFS Store

- **Purpose**: Binary distribution
- **Depends on**: Augmented `narinfo`

Instead of a "hybrid" store, where the narinfo index is served with HTTP but the data itself is served with IPFS, we can do an all-IPFS store with the data itself and mutable index stored in IPFS.
The Narinfo instead of being encoded the legacy line-oriented text format can be IPFS's native DAG-CBOR IPLD codec, which is like JSON + content address links (but stored as CBOR).
This allows Narinfos to reference each other and be nicely structured so the index is legible from Nix-agnostic IPFS tools and recursive pinning comes for free.

Read-only is easier, since IPFS data is immutable but "writable" stores are supported by simple printing back a new CID for the new store root after some modifications, or modifying a mutable IPNS reference.
IPNS is historically slow, but the update is automatic.

> Printing out a new CID for the index root allows the store administrator to update an out-of-bound mutable reference, but this cannot be automated because Nix doesn't know what the out-of-band method is.

Reviewer 2: > but this cannot be automated because Nix doesn't know what the out-of-band method is.

I'm not sure the point here. Just as updating IPNS can be automated we can automate publishing DNSLink entries for popular providers or any number of other methods. Maybe you trying to point out that this would be a protocol outside of IPFS?

Author: Yes exactly. It is quite possible to get Nix to call some hook, but there are perhaps details to figure out there (though I suppose we've rather wantonly added hooks so far...) and I didn't do it yet.

There is an interesting thing going on where I am very keen on the source code use-case, which due to it's statelessness is just much easier / more ergonomic on both the "provider" and "consumer" sides, but it also requires more up-front changes to Nix which could be controversial.

I look forward to RFC meetings to sort out all that.

Reviewer 4: It isn't as trendy, but Bittorrent's [ed25519-signed DHT entries](https://www.bittorrent.org/beps/bep_0044.html) can be used for this.  The protocol is very lightweight and easy to understand -- not to mention battle-tested.  Getting block-level content addressing requires [v2 torrents](http://www.bittorrent.org/beps/bep_0052.html) which are still not as popular as the v1 torrents -- but they share the same DHT network and uTP protocol.

Author: Yeah I am down to experiment the other protocols here. As I've long maintain, very little of the work here is *actually* specific to IPFS.

Author: Collapsing IPFS threads.

## Git file hashing

- **Purpose**: Source distribution and archival

In addition to the various forms of content-addressing Nix supports today ("text", "fixed" with either "flat" or "nar" serialization of file system objects), Nix should support Git hashing.
This support entails two basic things:

 - Content addresses are used to compute store paths.
 - Content addresses are used to verify store object integrity.

Git hashing would not support references (since references in Nix's sense are not a Git concept), but that is not an issue for the intended use-case of exchanging source code.

## Git file hashing for `buitins.fetch*`

- **Purpose**: Source distribution and archival
- **Depends on**: Git file hashing,

The builtin fetchers can also be made to work with git file hashing just as they support the other types.
In addition, Git repo fetching can leverage this better to than the other formats since the data in git repos is already content-addressed in this way.

## Content address or store path in Store interface

- **Purpose**: All distribution

Modify many store interface methods that today take store paths to instead accept *either* a store path or a content address.

> For stores that are purpose-built for Nix, like the ones we support today, all addressing can be done store paths, so the current interface is fine.

Author: Thanks! My thinking on what the store is / ought to be in general was definitely informed by working on this stuff :)

But for Nix-agnostic stores, store paths are rather useless as a key type because Nix-agnostic tools don't know about them.
They can, however, understand content addresses.
And from such a content address we can always produce a store path again, so there is no loss of functionality with existing stores.

## NAR info or content address normative in `ValidPathInfo`

- **Purpose**: Source distribution and archival
- **Depends on**: Content address or store path in Store interface,

As described in the first step, currently `NarHash` and `NarSize` are the *normative* fields which are used to verify a store object.
But if the store object is content-addressed, we don't need these, because the content address (`CA` field) will also suffice, all by itself.

> Relax the abstract `ValidPathInfo` type to merely require that *one of* `NarHash` and `NarSize` or `CA` be defined.

Reviewer 2: ```suggestion
Relax the abstract `ValidPathInfo` type to merely require that *one of* (`NarHash` and `NarSize`) or `CA` be defined.
```

I am mixing English and math syntax but I think this makes it more clear. Or maybe something like "both NarHash and NarSize or just CA be defined".

Author: ```suggestion
Relax the abstract `ValidPathInfo` type to merely require that *either* the pair of `NarHash` and `NarSize` or `CA` alone be defined.
```

Only content-addressed data is looked up, and it doesn't need any nar-info metadata before the data is all there.

We need the previous step for querying without fetching any data.
In that case since there is no narinfo index we're looking up, we don't get any additional metadata back.
But the content address key a successful query used is enough to create a bare-bones `ValidPathInfo` with a `CA` field, which with the enough step is valid.

(A bare-bones `ValidPathInfo` might sound sub-par, but for plain old content-addressed data it is fine.
Most of the other metadata in `ValidPathInfo` is really just for input-addressed derivation outputs, and is thus obviated by CA derivation trust maps which contain the same data but more naturally.)

## Wrapped git objects with references

- **Purpose**: Binary distribution
- **Depends on**: Git file hashing

Merkelized formats like git file hashing are better than NAR because that allow for very natural deduplication and minimal transfers.
This is the same benefit we get today with Nix within a closure of multiple store objects, now also *within a single store object*.

> But git has no notion of Nix-style references, so plain git hashing is only suitable for leaf store objects without references (like source code).

Author: Yes, but I wouldn't want to "hack in" our references as invalid submodules.

Our references are also not associated with a single "place" in the file system tree. They are really store object wide, or if one thinks of scanning associated with hashes in specific files. Neither of those corresponds to where submodules are referenced.
